Custom basis for 3D HCurl and HDiv in 1D or 2D [lo-dim-em-dev]

mlstowell opened the pull request:
Adding ND and RT basis functions for 3D vector fields on 1D or 2D elements.  These are intended for modeling electromagnetic fields on 1D or 2D domains in the presence of anisotropic material properties.   In such cases the in-domain and out-of-domain polarizations can mix.  These basis functions avoid a great deal of code complexity needed to handle this type of mixing.

Anisotropic materials can arise in wave propagation through crystals or magnetized plasmas.

This PR adds three examples analogous to examples ex3p, ex6p, and ex13p but employing the new basis functions. Serial versions of two of these could also be added if desired (the third is an eigenvalue example which would remain parallel).

To facilitate the use of these basis functions I have added two new methods to the `FiniteElement` base class which specify the dimensions of the vector field and its curl.  Previously we made the assumption that `vdim == dim` or `vdim == sdim` but this is not the case for these bases.  Also the curl could have dimension 1 or 3 in 2D.  We could make assumptions about the dimension of the curl but I felt it was cleaner to declare it in a more obvious manner.

Although this PR adds many lines of code the vast majority of these lines are in unit tests so the library itself is not significantly larger.

Background:
I'm sure we're all aware that modern electromagnetic modeling is often accomplished with Nedelec and Raviart-Thomas finite elements (or an equivalent Yee lattice when finite differences will suffice). Unlike position or velocity fields which map `R^d` to `R^d` where `d`, clearly, is the dimension of the space, electromagnetic fields are inherently three dimensional. In other words the electric and magnetic vector fields always map `R^d` to `R^3`.

There are, of course, many special cases where symmetries of the problem ensure that one or more components of the electric or magnetic field will be zero.  For example, if the problem is invariant with respect to translations in the z direction, and the materials are isotropic, the electric field component in the z direction might be guaranteed to be zero and the standard Nedelec basis functions defined on triangles and/or quadrilaterals would suffice to describe such a field. Alternatively, the x and y components of the electric field might be guaranteed to be zero and the z component of the field could be accurately described by an H1 scalar finite element on triangles and/or quadrilaterals. In some cases none of the field components are guaranteed to be zero but the symmetries of the problem ensure that the x and y components are decoupled from the z component. In these cases the full electric field is often computed by two solving separate PDEs; one for the x and y components, and one for the z component, and combining the results during post-processing if at all.

Such special cases only exist if the material coefficients possess symmetries which are compatible with the translational or rotational symmetries of the problem domain, boundary conditions, and sources. In crystallography and plasma physics the material coefficients are generally anisotropic and not aligned with the other symmetries. This prevents the decoupling of the field components.

Motivation:
In order to model electromagnetic fields in reduced dimensions with anisotropic materials in MFEM we need to build `BlockOperators` and `BlockVectors` to describe our coupled system of PDEs. This can be done with the existing operators but it is far from convenient and there are many places where user error can lead to incorrect systems of PDEs, boundary conditions, or source functions.

Using `BlockVectors` in this manner is a bit unnatural. Building block systems to simultaneously solve for say density and velocity in the Darcy problem makes perfect sense because these are separate fields. For the electromagnetic fields this splitting into `BlockVectors` is inconvenient and unnatural. It would seem to be more logical to keep E_xy and E_z tightly coupled in a single `GridFunction` so that its curl, energy, etc. could be easily computed in a single operation.

The alternative implemented in this PR allows a user to setup the same linear system that would be used in a 3D simulation using the same integrators. This is accomplished by creating a set of basis functions which, in a sense, exist in between the standard 3D and 2D basis functions. We can think of the standard 2D Nedelec and Raviart-Thomas basis functions as being obtained from their 3D counterparts by first restricting the domain of the 3D basis functions to 2D and then projecting the vector-valued basis functions onto the 2D plane. This pair of operations eliminates many of the 3D basis functions by reducing effectively them to zero. The basis functions introduced in this PR perform the Restriction to 2D, hence the label `R2D`, but not the projection. Consequently they retain basis functions which are directed in the z direction.

Another way to look at the new Nedelec basis functions is that the `ND_R2D` basis functions are the direct sum of the standard 2D Nedelec basis functions and the 2D H1 basis functions multiplied by the unit vector in the z direction. Similarly the `RT_R2D` basis functions are a direct sum of 2D RT basis functions and 2D L2 basis functions (with map type INTEGRAL) multiplied by the z directed unit vector. The new `R1D` basis functions can be similarly constructed using two copies of the H1 or L2 basis functions for the y and z components. From this point of view they indeed look strange but I feel that the "restriction" view point described above is quite natural.

I think it's rather impressive that so much of the existing MFEM machinery works seamlessly with these new basis functions. The changes needed in some of the integrators were really quite small and, I think, make those integrators more robust.

<!--GHEX{"id":1989,"author":"mlstowell","editor":"v-dobrev","reviewers":["brendankeith","dylan-copeland"],"assignment":"2021-04-14T07:49:51-07:00","approval":"2022-03-09T07:51:37.105Z","merge":"2022-03-11T19:58:24.803Z"}XEHG--><!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1989](https://github.com/mfem/mfem/pull/1989) | @mlstowell | @v-dobrev | @brendankeith + @dylan-copeland | 04/14/21 | 03/08/22 | 03/11/22 | |
<!--ELBATXEHG-->

v-dobrev commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

v-dobrev commented:
@mlstowell, can you provide some background on these new finite elements, maybe a paper reference? I tried to look at the implementation but it all seems quite strange and non-standard.

dylan-copeland commented on fem/bilininteg.cpp:
```diff
@@ -526,16 +527,58 @@ void MixedVectorIntegrator::AssembleElementMatrix2(
-            test_shape_tmp(j,2) = test_shape(j,0) * V(1) -
-                                  test_shape(j,1) * V(0);
+            // Compute shape_tmp = test_shape x V
+            // V will always be of length 3
```
Should there be `MFEM_VERIFY(VQ->GetVDim() == 3, "")`?

mlstowell replied:
Good point

dylan-copeland commented on fem/fe.hpp:
```diff
@@ -3282,6 +3293,427 @@ class ND_SegmentElement : public VectorFiniteElement
+                            ElementTransformation &Trans,
+                            DenseMatrix &curl) const;
+   /*
+    virtual void GetLocalInterpolation(ElementTransformation &Trans,
```
Should the commented code here be kept?

dylan-copeland commented on fem/fe.cpp:
```diff
@@ -12627,6 +12637,2047 @@ void ND_SegmentElement::CalcVShape(const IntegrationPoint &ip,
+
+   shape(1,0) = 0.0;
+   shape(1,1) = 1.0;
+}
```
According to the documentation, this is an arbitrary order element that is constant in the second and third dimension. Why is just a constant 2D basis of dimension 2?

mlstowell replied:
It looks weird, doesn't it? This finite element is analogous to the `PointFiniteElement` which appears alongside the `H1_FiniteElements`. It's "arbitrary order" in the sense that it is compatible with 1D Nedelec finite elements of any order. In the constructor I pass an integer `p` even though this is not really relevant which is the "order" used in the `ND_R1D_FECollection`. Following the discussion in PR #2163 I'm really not sure what to do with these integers.

dylan-copeland replied:
This is the first time I've looked at finite elements at a point, and it doesn't make sense to me yet. I see that `H1_FECollection` has `PointFiniteElement`, and it seems to be just 0-order (constant). How is it actually used, and why? Does it just get used for evaluating a point value of a `GridFunction` or something, so it only needs to be a constant that gets multiplied by a value given by a coefficient?

mlstowell replied:
The only uses I can think of are as boundary elements. For example, when preparing a Neumann BC for a PDE on a 1D domain. This involves an integral over the boundary which in this case is just evaluating the BC function and the basis function at a single point. Rather than treating 0D boundaries as special cases we have 0D basis functions which can be plugged into the usual machinery and produce the proper results.

dylan-copeland replied:
Then maybe the source of confusion is the comment in the hpp file that the elements are "arbitrary order." The documentation for `PointFiniteElement` does not say anything about the order. Would it be more accurate to say it is 0-order and works for 1D boundary evaluations with arbitrary order 1D elements? Or is it better just not to mention order, since it can be confusing without a complete explanation of how these elements are used?

mlstowell replied:
Agreed. I rewrote the comment to remove "order" and hopefully summarized its purpose a little better.

dylan-copeland commented on fem/fe.hpp:
```diff
@@ -3282,6 +3293,427 @@ class ND_SegmentElement : public VectorFiniteElement
+/** ND_R1D_SegmentElement provides a representation of a 3D Nedelec
+    basis where the vector field is assumed constant in the second and
+    third dimensions.
+*/
```
The basis functions vary in all 3 components. Would it be clearer to say that the vector field is three-dimensional and depends only on the first dimension of position?

dylan-copeland commented on fem/fe.cpp:
```diff
@@ -12627,6 +12637,2047 @@ void ND_SegmentElement::CalcVShape(const IntegrationPoint &ip,
+      }
+}
+
+
```
This file is getting very long. Should the new code be put in a different `cpp` file?

mlstowell replied:
I would support splitting `fe.cpp` and `fe.hpp` into multiple files. Perhaps one for H1, one for ND, one for RT, one for L2, and keep common base classes in `fe.hpp` and `fe.cpp`. Other basis functions could either stay in the current files or be placed in a miscellaneous file.

dylan-copeland replied:
That sounds good to me. We already have similar splittings of the `bilininteg_*` files.

dylan-copeland commented on fem/fe_coll.cpp:
```diff
@@ -2751,6 +2751,451 @@ ND_Trace_FECollection::ND_Trace_FECollection(const int p, const int dim,
+      delete ND_Elements[g];
+   }
+}
+
```
Extra line

mlstowell replied:
That's a matter of opinion, I suppose. This file often uses a double blank line between classes. See lines 1531, 1918, 1938, etc..

dylan-copeland commented on fem/gridfunc.cpp:
```diff
@@ -328,7 +328,28 @@ int GridFunction::VectorDim() const
+   }
+   if (!fe || fe->GetRangeType() == FiniteElement::SCALAR)
+   {
+      return 2 * fes->GetMesh()->SpaceDimension() - 3;
```
Is this correct in 1D, returning -1?

mlstowell replied:
I'm not sure it needs to be correct in 1D but I could be mistaken. This `if` statements handles the case where we have multiple copies of a scalar field and our current assumptions are that such a field is either 2 or 3 dimensional. I believe a 1D scalar field, with vdim=1, would throw an error before calling this method.

Of course, it wouldn't hurt to simplify this logic and maybe even document this method so that we know what to expect from it...

dylan-copeland commented on tests/unit/fem/test_r1d_bilininteg.cpp:
```diff
@@ -0,0 +1,4753 @@
+            REQUIRE( df_l2.ComputeL2Error(divF3_coef) < tol );
+         }
+         /*
+              SECTION("Divergence of RT")
```
Why is this commented?

mlstowell replied:
Good question. I don't see a reason so keep it.

dylan-copeland commented on tests/unit/fem/test_r2d_bilininteg.cpp:
```diff
@@ -0,0 +1,4671 @@
+            REQUIRE( df_l2.ComputeL2Error(divF3_coef) < tol );
+         }
+         /*
+              SECTION("Divergence of RT")
```
Another commented section?

mlstowell replied:
Hmm... I'll take it out

dylan-copeland commented on tests/unit/fem/test_r2d_bilininteg.cpp:
```diff
@@ -0,0 +1,4671 @@
+   VectorFunctionCoefficient DF3_coef(vdim, DF3);
+   VectorFunctionCoefficient MF3_coef(vdim, MF3);
+
+   for (int type = (int)Element::TRIANGLE;
```
It looks like the files `test_r1d_bilininteg.cpp` and `test_r2d_bilininteg.cpp` are similar enough that they could be combined into one file, with a loop over all types of elements. Depending on the dimension, a pointer to a `FiniteElementCollection` could be set to a new 1D or 2D collection. Maybe there are complications I don't see, but if this could be done, it would save 4600+ lines of code.

mlstowell replied:
I suspected you might comment on this :smirk:

It would probably be possible. One issue would be the tests which mix the new classes with derivatives of the preexisting classes such as gradient of an H1. These tend to involve multiplication with rectangular matrices like 3x1 or 3x2 mapping matrices. The R1D versions tend to have more pieces than the R2D versions which may not be trivial to merge. Another issue is the analytic functions at the top of the page. These are generally named the same in the two files but they have different definitions.  Renaming and merging these would not be fun although it could certainly be done.

The last issue that comes to mind might be the most important for me. In the current implementation the `TEST_CASE` statements can be used to run subsets of the unit tests and this is _so_ convenient when tracking down problems. If we merged these unit tests into a single file we'd need to merge the names appearing in these `TEST_CASE` statements and we would be able to just run R1D or R2D cases separately. This issues has already bit me after we merged other tests and I've had to temporarily edit files so that they skip certain tests so that I can focus on more urgent things. Clearly there are workarounds for this issue and I could manage if the size of the unit test files is something we need to reduce.

In short, I am willing to merge these but I'd rather not.

dylan-copeland replied:
It sounds reasonable to keep them separate then. I think we should merge unless there are good reasons not to, and you provided some good reasons.

mlstowell replied:
That's a relief 😌. Thanks, @dylan-copeland !

dylan-copeland commented on examples/ex29p_proposed.cpp:
```diff
@@ -0,0 +1,540 @@
+//               discretization error when the exact solution is known. These
+//               restricted spaces allow the solution of 1D or 2D
+//               electromagnetic problems which involve 3D field vectors.  Such
+//               problems arise in plasma physics and crystalography.
```
crystallography

mlstowell replied:
Good catch

dylan-copeland commented on fem/fe.hpp:
```diff
@@ -301,13 +303,21 @@ class FiniteElement
+   int GetVDim() const { return vdim; }
+
+   /// Returns the dimension of the curl for vector-valued finite elements
+   int GetCDim() const { return cdim; }
```
`GetCurlDim` would be more readable, but I don't have a strong preference.

mlstowell replied:
I agree and I also don't have a strong preference. Let's see if anyone else has a preference. I'll be happy to change it if that's the consensus.

brendankeith replied:
I prefer `GetCurlDim`

dylan-copeland commented on examples/ex29p_proposed.cpp:
```diff
@@ -0,0 +1,540 @@
+   DenseMatrix sigmaMat(3);
+   sigmaMat(0,0) = 2.0; sigmaMat(1,1) = 2.0; sigmaMat(2,2) = 2.0;
+   sigmaMat(0,2) = 0.0; sigmaMat(2,0) = 0.0;
+   sigmaMat(0,1) = M_SQRT1_2; sigmaMat(1,0) = M_SQRT1_2;
```
This is the first time I ever saw constants like `M_SQRT1_2`, but Google helped me learn what they are. Are other people usually familiar with these, or should there be a comment about the definition?

mlstowell replied:
This may be one of the more obscure ones. I can add a comment at the end of the line if that would help to clarify things.

dylan-copeland replied:
Yes, it can be a short comment like `// 1/sqrt(2) in cmath`

mlstowell commented:
> @mlstowell, can you provide some background on these new finite elements, maybe a paper reference? I tried to look at the implementation but it all seems quite strange and non-standard.

Sorry for the delayed response. I missed this question until @dylan-copeland pointed it out this morning.

I am not aware of any papers that describe these finite elements and a quick literature search didn't turn up any. These finite elements arose out of a few short discussions that I had with @homijan in the distant past i.e. before the pandemic. I would be very interested to hear his feedback on this PR as well.

I'll add more background and motivation to the comments at the start of this PR so that they'll be easy to find.

mlstowell commented on fem/fe.cpp:
```diff
@@ -22,10 +22,12 @@ namespace mfem
    : Nodes(Do)
 {
    dim = D ; geom_type = G ; dof = Do ; order = O ; func_space = F;
+   vdim = VD ; cdim = CD;
```
I agree but I'm following the style set in the previous line. I can change both lines if people agree but in this case I think the spaces do make it more readable.

brendankeith commented:
> > @mlstowell, can you provide some background on these new finite elements, maybe a paper reference? I tried to look at the implementation but it all seems quite strange and non-standard.
> 
> Sorry for the delayed response. I missed this question until @dylan-copeland pointed it out this morning.
> 
> I am not aware of any papers that describe these finite elements and a quick literature search didn't turn up any. These finite elements arose out of a few short discussions that I had with @homijan in the distant past i.e. before the pandemic. I would be very interested to hear his feedback on this PR as well.
> 
> I'll add more background and motivation to the comments at the start of this PR so that they'll be easy to find.

The new background and motivation is quite nice! :)

brendankeith commented on examples/ex30p_proposed.cpp:
```diff
@@ -0,0 +1,532 @@
+//
+// Compile with: make ex30p
+//
+// Sample runs:  mpirun -np 4 ex30p -m ../data/inline-segment.mesh
```
`❯ ./ex30p_proposed` runs fine. However, when I execute the first suggested sample run, `❯ ./ex30p_proposed -m ../data/inline-segment.mesh`, I receive a segmentation fault. The segmentation also appears when running `❯ mpirun -np 4 ex30p_proposed -m ../data/inline-segment.mesh`. I have not checked the remaining sample runs in this file.

mlstowell replied:
Actually I had to tweak HYPRE's implementation of AMS so that it would support 1D problems. I had forgotten about this. I have a pull request ([#269](https://github.com/hypre-space/hypre/pull/269)) with HYPRE which I should really followup on.

dylan-copeland replied:
Nice catch, I didn't try this sample run. @mlstowell should this sample run be removed until the hypre PR is merged, since otherwise the MFEM testing will fail?

mlstowell replied:
I have removed the 1D sample runs.

brendankeith commented on examples/ex32p_proposed.cpp:
```diff
@@ -0,0 +1,455 @@
+//
+// Compile with: make ex32p
+//
+// Sample runs:  mpirun -np 4 ex32p -m ../data/square-disc.mesh -o 1
```
More than one of sample runs have visualization problems, similar to the case reported above, when they are run with MPI.

It appears that all of the sample runs work when run in serial. However, some of the runs deliver very suspicious refinement patterns. It is possible that the error estimator being used here is not appropriate "right out of the box" for practical applications. Are you sure that this example should be merged into the main code? Another option is stating somewhere that there is "experimental support" for AMR/a posteriori error estimation.

mlstowell replied:
Yes, I noticed this a few days ago and I've been trying to track down the problem (without much luck). I may pull `ex32p_proposed ` out of this PR. Sorry for the hassle.

psocratis replied:
> Yes, I noticed this a few days ago and I've been trying to track down the problem (without much luck). I may pull `ex32p_proposed ` out of this PR. Sorry for the hassle.

I was playing with this too, just now, and what happens in parallel runs is that the solver produces NANs, in a non-deterministic way. I noticed that the source of this is the call to 
```cpp
sol.ProjectBdrCoefficientTangent(E, ess_bdr);
````
If replaced with `sol.ProjectCoefficient(E);` the non-deterministic behavior goes away. There is still a Bad termination at the 17th AMR iteration but it's deterministic and up to that point everything looks good.

mlstowell replied:
This sounds like a useful clue. Thanks, @psocratis!

Ever since these `ProjectBdr*` methods started using `AccumulateAndCount*` they keep causing me trouble. We should see if we have any unit tests that cover these. Not in this PR but soon...

mlstowell replied:
I created an issue (#2226) so that we don't forget to resolve this.

brendankeith commented on examples/ex31p_proposed.cpp:
```diff
@@ -0,0 +1,531 @@
+//
+// Compile with: make ex31p
+//
+// Sample runs:  mpirun -np 4 ex31p -m ../data/inline-segment.mesh
```
At this point, it's clear why this failed. I'm setting this as "resolved."

brendankeith commented on examples/ex32p_proposed.cpp:
```diff
@@ -0,0 +1,455 @@
+   Mesh *mesh = new Mesh(mesh_file, 1, 1);
+   dim = mesh->Dimension();
+
+   MFEM_VERIFY(dim == 2, "This exmaple requires a 2D mesh.");
```
This **example** requires...

mlstowell commented:
I've removed the 1D sample runs to await possible support in AMS.

I've also moved the AMR example to a separate PR since it's not working in parallel for some unknown reason.

mlstowell commented:
Added a serial version of ex30 (and this one works on 1D meshes since it doesn't rely on HYPRE)

brendankeith commented on examples/ex30_proposed.cpp:
```diff
@@ -0,0 +1,466 @@
+// Sample runs:  ex30 -m ../data/inline-segment.mesh -o 2
+//               ex30 -m ../data/hexagon.mesh -o 2
+//               ex30 -m ../data/star.mesh -o 2
+//               ex30 -m ../data/fichera.mesh -o 3 -r 1
```
Although the `../data/amr-hex.mesh -r 1` and `ex30 -m ../data/amr-hex.mesh -r 1` execute without errors, I am not sure whether they should be here. I thought that this basis was meant for 1D and 2D domains. On that note, a sample run where the domain is an embedded 2D surface may be interesting. Thoughts?

brendankeith replied:
@mlstowell Any thoughts on this?

brendankeith commented on examples/ex31p_proposed.cpp:
```diff
@@ -0,0 +1,530 @@
+//
+// Sample runs:  mpirun -np 4 ex31p -m ../data/hexagon.mesh -o 2
+//               mpirun -np 4 ex31p -m ../data/star.mesh
+//               mpirun -np 4 ex31p -m ../data/square-disc.mesh -o 2 -n 4 -rs 1
```
Suspicious results when running `mpirun -np 4 ex31p_proposed -m ../data/square-disc.mesh -o 2 -n 4 -rs 1`
<img width="802" alt="Screen Shot 2021-05-06 at 12 01 10 PM" src="https://user-images.githubusercontent.com/11756601/117351580-c0d2ca80-ae62-11eb-8b57-582873a217be.png">

brendankeith replied:
Is the genus of the domain being properly handled here? The cohomology is non-trivial here....

brendankeith replied:
> Not that different results are produced from `mpirun -np 1 ex31p_proposed -m ../data/square-disc.mesh -o 2 -n 4 -rs 1`

This also could be related to https://github.com/mfem/mfem/issues/2226. Thanks @psocratis 
<img width="799" alt="Screen Shot 2021-05-06 at 12 56 56 PM" src="https://user-images.githubusercontent.com/11756601/117358052-8c630c80-ae6a-11eb-87d5-c5cc5330e8b4.png">

mlstowell replied:
![GLVis_s01](https://user-images.githubusercontent.com/6109571/117361494-f4b3ed00-ae6e-11eb-9ae5-8557fd3afe75.png)
This result looks much better on the NURBs mesh. In any case this is certainly a good question. HypreAME should be using the discrete gradient operator to weed out gradients of scalar fields. This should leave the rotational fields we're interested in and a single irrotational field due to the hole in the topology. This irrotational field doesn't wrap around the hole as I would expect but I cannot say if it's correct or not.

As for the other three panels; they look ugly. I was considering rescaling the panels so that the XY panel and the Z panel use the same color bar. I think this would help to alleviate some of the confusion caused by this set of images.

Do you think this rescaling is a good idea in general? Should we avoid this topology in particular? Should we switch to the NURBs mesh (I think we should)?

brendankeith replied:
It's really interesting that the results are so much poorer with the high order mesh than with the NURBS mesh. I would suggest that the default behaviour be not to plot the zero eigenmodes. That would bypass the need for any rescaling.

brendankeith replied:
It may be good to keep this there, alongside the NURBS mesh example. There is some implicit information in there. If someone runs both, they may decide right away that it's worth using a NURBS description of their domain geometry.

brendankeith replied:
> ![GLVis_s01](https://user-images.githubusercontent.com/6109571/117361494-f4b3ed00-ae6e-11eb-9ae5-8557fd3afe75.png)
> This result looks much better on the NURBs mesh. In any case this is certainly a good question. HypreAME should be using the discrete gradient operator to weed out gradients of scalar fields. This should leave the rotational fields we're interested in and a single irrotational field due to the hole in the topology. This irrotational field doesn't wrap around the hole as I would expect but I cannot say if it's correct or not.
> 
> As for the other three panels; they look ugly. I was considering rescaling the panels so that the XY panel and the Z panel use the same color bar. I think this would help to alleviate some of the confusion caused by this set of images.
> 
> Do you think this rescaling is a good idea in general? Should we avoid this topology in particular? Should we switch to the NURBs mesh (I think we should)?

@mlstowell Can you share the command you used to produce this figure? I am not seeing such good results with `../data/square-disc-nurbs.mesh` anymore :/

brendankeith replied:
For example, `❯ mpirun -np 4 ex31p_proposed -m ../data/square-disc-nurbs.mesh -o 2 -n 4 -rs 1` produces
<img width="244" alt="Screen Shot 2021-05-18 at 10 53 20 AM" src="https://user-images.githubusercontent.com/11756601/118700208-4477aa00-b7c7-11eb-9421-6698a4fc1a65.png">

mlstowell replied:
I don't recall exactly but `./ex31p_proposed -m ../data/square-disc-nurbs.mesh -rs 4 -o 2` looks quite similar. This one's not bad either `./ex31p_proposed -m ../data/square-disc-nurbs.mesh -rs 3 -o 3`.

brendankeith replied:
OK. Perhaps the default should be `./ex31p_proposed -m ../data/square-disc-nurbs.mesh -rs 3 -o 3`
`mpirun -np 4 ex31p_proposed -m ../data/square-disc-nurbs.mesh` doesn't look very nice

brendankeith commented on examples/ex31p_proposed.cpp:
```diff
@@ -0,0 +1,530 @@
+//
+// Compile with: make ex31p
+//
+// Sample runs:  mpirun -np 4 ex31p -m ../data/hexagon.mesh -o 2
```
The eigenmodes are very pretty :star-struck:, however, they do not appear to preserve the symmetries of the domain. Could this be a visualization issue or a convergence issue?  Or is it possibly something else?
<img width="802" alt="Screen Shot 2021-05-06 at 12 12 21 PM" src="https://user-images.githubusercontent.com/11756601/117352848-515dda80-ae64-11eb-9694-fedff22710e3.png">
<img width="801" alt="Screen Shot 2021-05-06 at 12 11 34 PM" src="https://user-images.githubusercontent.com/11756601/117352898-5d499c80-ae64-11eb-889f-4cb958990f10.png">

I'm not sure whether there should be repeated eigenvalues with this type of problem. @mlstowell do you know?

mlstowell replied:
This problem uses an anisotropic material tensor in the mass matrix. The strongest eigenvalue of this tensor is directed in the (1,1,1) direction, the next in the (0,0,1) direction, and the smallest in the (1,-1,1) direction. This is why the modes tend to be aligned with +/-45 degrees.

mlstowell commented:
@v-dobrev, et al, Dylan was concerned that the `fe.cpp` and `fe.hpp` files are becoming too long and I tend to agree. I'd like to propose the following split:

- `fe.[ch]pp` to contain the `FiniteElement` base class and any other common definitions or code
- `fe_h1.[ch]pp` to contain the H1 basis functions
- `fe_l2.[ch]pp` to contain the L2 basis functions
- `fe_nd.[ch]pp` to contain the ND basis functions
- `fe_rt.[ch]pp` to contain the RT basis functions 
- `fe_misc.[ch]pp` to contain other classes that don't fit naturally in the above files

This is probably the start of a longer discussion but the real question here is should this splitting (or something similar) be done in this PR?

tzanio commented:
I think it is time to split the `fe` files, there is a lot there and it may be difficult to find to separate the important from less frequently used stuff.

v-dobrev commented:
> This is probably the start of a longer discussion but the real question here is should this splitting (or something similar) be done in this PR?

I'd prefer to do this in a separate PR.

homijan commented:
@mlstowell I think that adding the `RT_R1D`/`RT_R2D` examples is unnecessary it it holds back this PR. Adding it in a future PR makes more sense.

mlstowell commented:
@v-dobrev, I think this now incorporates all the changes we discussed the other day. Thanks again for the useful feedback!

v-dobrev commented on fem/fe/fe_base.hpp:
```diff
@@ -360,7 +368,8 @@ class FiniteElement
    /** @brief Evaluate the values of all shape functions of a scalar finite
        element in physical space at the point described by @a Trans. */
    /** The size (#dof) of the result Vector @a shape must be set in advance. */
-   void CalcPhysShape(ElementTransformation &Trans, Vector &shape) const;
+   virtual void CalcPhysShape(ElementTransformation &Trans,
+                              Vector &shape) const;
```
Why make this method virtual? To me it looks like it does not need to be.

mlstowell replied:
I don't think it makes a difference for this PR. I can change it back. I probably just didn't see a reason why it shouldn't be virtual.

v-dobrev commented:
@mlstowell,

can you also:
- [ ] Update CHANGELOG
- [ ] List the new example in doc/CodeDocumentation.dox

Also, I'll open an issue to document the new examples in the MFEM webpage repo, https://github.com/mfem/web.

mlstowell commented on fem/bilininteg.cpp:
```diff
@@ -600,17 +664,17 @@ void MixedScalarVectorIntegrator::AssembleElementMatrix2(
    int trial_nd = trial_fe.GetDof(), test_nd = test_fe.GetDof(), i;
    int sca_nd = sca_fe->GetDof();
    int vec_nd = vec_fe->GetDof();
-   int spaceDim = Trans.GetSpaceDim();
+   int vdim = GetVDim(*vec_fe);
```
This is a tricky one. In this case `GetVDim` is a member function of the integrator which is meant to be overridden by derived classes. So, we don't want to supersede choices made in the derived classes. Perhaps the best resolution is to store the space dimension as member data so that we can compute the `max` inside `GetVDim` by default but still allow derived classes to make their own determination.

v-dobrev commented:
Merged in `next` for testing...

adam-sim-dev commented:
Dear Mark, @mlstowell , I have a question about the ND_R2D element. Does it support the 3D source? For example, a horizontal electric dipole along the x-axis in the xy plane, or a vertical magnetic dipole on the xy plane. I do see in ex31, the RHS is 3D.

adam-sim-dev commented:
Another question, are the three dimensions of the shape functions in ND_R2D_TriangleElement::CalcVShape and ND_R2D_TriangleElement::CalcCurlShape corresponding to x,y, and z, where x, y are the two dimensions of the 2D model?
